### DISCUSSION TITLE
FluxCD Update data partially on existing ConfigMap

### DISCUSSION THREAD

**Discussion by jgilnom-san** [fluxcd/flux2#2490] [Q&A] [answered]
Hi all.

We have fluxcd installed on EKS (latest version of flux) and we need to do the following. EKS adds by default, for IAM authentication, the configmap named  aws-auth in kube-system.

This configmap has 2 sections in the data:

---
apiVersion: v1
kind: ConfigMap
metadata:
  name: aws-auth
  namespace: kube-system
data: 
  mapRoles: |
    xxxxxxxxxxxxxxxxxxx
mapUsers: |
    xxxxxxxxxxxxxxxxxxx

the part that I would like to manage with fluxcd is the mapUsers so. Could I manage only that part of the configmpap using fluxcd and let mapRoles managed by the own EKS? 

It would be something like "incremental modification" or "partial modification". Don't know if with something like configmapgenerator or with some kustomization configuration it could be achieved or not.

Thanks in advance!!

**Comment by stefanprodan** [comment]
Can you please post here the managed fields of that object `kubectl get cm aws-auth —show-managed-fields`

**Comment by jgilnom-san** [comment]
of course:

  managedFields:
  - apiVersion: v1
    fieldsType: FieldsV1
    fieldsV1:
      f:data:
        .: {}
        f:mapRoles: {}
    manager: vpcLambda
    operation: Update
    time: 2022-02-27T09:30:12Z
  - apiVersion: v1
    fieldsType: FieldsV1
    fieldsV1:
      f:data:
        f:mapUsers: {}
    manager: node-fetch
    operation: Update
    time: 2022-02-28T17:53:02Z

**Reply by stefanprodan** [reply]
This looks Ok, you can place the ConfigMap in Git, add the extra keys there and Flux will merge them at apply time. Make sure `mapRoles` and `mapUsers` are not in Git.

**Reply by stefanprodan** [reply]
You should also disable pruning for this CM

**Reply by jgilnom-san** [reply]
ok will try! many thanks!

**Comment by jgilnom-san** [comment, marked as answer]
@stefanprodan I tried this but with no success:

**aws-auth.yaml:**
apiVersion: kustomize.toolkit.fluxcd.io/v1beta1
kind: Kustomization
metadata:
  name: aws-auth
  namespace: flux-system
spec:
  interval: 10m0s
  sourceRef:
    kind: GitRepository
    name: flux-system
  path: ./infrastructure/base/aws-auth
  prune: false
  validation: client
  postBuild:
    substitute:
      account: xxxxxxxxx


In the git folder ./infrastructure/base/aws-auth:

**kustomization.yaml:**
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
namespace: default
resources:
  - configmap.yaml

**configmap.yaml:**
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: aws-auth
  namespace: default
data:
  mapWhatever: |
    - userarn: arn:aws:iam::${account}:user/username
      username: github-ci
      groups:
        - "groupname"


but when the kustomization is applied, the original configmap aws-auth that contained mapUsers and mapRoles was replaced with a configmap with just mapWhatever. What am i doing wrong or what I missed?

Thanks.

**Reply by stefanprodan** [reply]
The aws-auth ConfigMap is in kube-system, why are you setting it to default?

**Reply by jgilnom-san** [reply]
yes is for testing if it was working or not and don't affect the current configmap that is in use.

**Reply by stefanprodan** [reply]
So did you applied the ConfigMap using the the AWS manager??

**Reply by jgilnom-san** [reply]
yes manually applied a copy of the aws-auth configmap in default namespace.

**Reply by stefanprodan** [reply]
You need to apply it using manager vpcLambda

**Reply by stefanprodan** [reply]
Flux undoes all changes made with kubectl, but it will not undo vpcLambda.

**Reply by jgilnom-san** [reply]
ok, checking.

**Reply by stefanprodan** [reply]
See here how you can set a different manager https://fluxcd.io/docs/faq/#why-are-kubectl-edits-rolled-back-by-flux

**Reply by jgilnom-san** [reply]
thanks!

**Comment by tuxillo** [comment]
@jgilnom-san If I understood correctly you are maintaining the `aws-auth` configmap from the AWS manager instead of Flux because it is not doable with Flux?
